File: golang/lib/services/service_context.go
/*
 *    Copyright 2021 Kurtosis Technologies Inc.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 *
 */

package services

import (
	"context"
	"github.com/kurtosis-tech/kurtosis-engine-api-lib/golang/kurtosis_engine_rpc_api_bindings"
	"github.com/kurtosis-tech/kurtosis-engine-api-lib/golang/lib/binding_constructors"
	"github.com/palantir/stacktrace"
)

// Docs available at https://docs.kurtosistech.com/kurtosis-engine-api-lib/lib-documentation
type ServiceContext struct {
	client			kurtosis_engine_rpc_api_bindings.EngineServiceClient
	enclaveId       string
	serviceId		ServiceID
	ipAddress		string
	sharedDirectory	*SharedPath
}

func NewServiceContext(
		client kurtosis_engine_rpc_api_bindings.EngineServiceClient,
		enclaveId string,
		serviceId ServiceID,
		ipAddress string,
	    sharedDirectory *SharedPath) *ServiceContext {
	return &ServiceContext{
		client:          client,
		enclaveId:       enclaveId,
		serviceId:       serviceId,
		ipAddress:       ipAddress,
		sharedDirectory: sharedDirectory,
	}
}

// Docs available at https://docs.kurtosistech.com/kurtosis-engine-api-lib/lib-documentation
func (self *ServiceContext) GetServiceID() ServiceID {
	return self.serviceId
}

// Docs available at https://docs.kurtosistech.com/kurtosis-engine-api-lib/lib-documentation
func (self *ServiceContext) GetIPAddress() string {
	return self.ipAddress
}

// Docs available at https://docs.kurtosistech.com/kurtosis-engine-api-lib/lib-documentation
func (self *ServiceContext) GetSharedDirectory() *SharedPath {
	return self.sharedDirectory
}

// Docs available at https://docs.kurtosistech.com/kurtosis-engine-api-lib/lib-documentation
func (self *ServiceContext) ExecCommand(command []string) (int32, string, error) {
	serviceId := self.serviceId
	args := binding_constructors.NewExecCommandArgs(
		self.enclaveId,
		string(serviceId),
		command,
	)
	resp, err := self.client.ExecCommand(context.Background(), args)
	if err != nil {
		return 0, "", stacktrace.Propagate(
			err,
			"An error occurred executing command '%v' on service '%v'",
			command,
			serviceId)
	}
	return resp.ExitCode, resp.LogOutput, nil
}

File: typescript/src/lib/services/service_context.ts
import { EngineServiceClient } from '../../kurtosis_engine_rpc_api_bindings/engine_service_grpc_pb'; 
import { ExecCommandArgs, ExecCommandResponse } from '../../kurtosis_engine_rpc_api_bindings/engine_service_pb';
import { ServiceID} from './service';
import { SharedPath } from './shared_path';
import { newExecCommandArgs} from "../constructor_calls";
import { ok, err, Result } from 'neverthrow';
import * as grpc from "grpc";

// Docs available at https://docs.kurtosistech.com/kurtosis-engine-api-lib/lib-documentation
export class ServiceContext {
    
    private readonly client: EngineServiceClient;
    private readonly enclaveId: string;
    private readonly serviceId: ServiceID;
    private readonly ipAddress: string;
    private readonly sharedDirectory: SharedPath;

    constructor(
            client: EngineServiceClient,
            enclaveId: string,
            serviceId: ServiceID,
            ipAddress: string,
            sharedDirectory: SharedPath) {
        this.client = client;
        this.enclaveId = enclaveId;
        this.serviceId = serviceId;
        this.ipAddress = ipAddress;
        this.sharedDirectory = sharedDirectory;
    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-engine-api-lib/lib-documentation
    public getServiceID(): ServiceID { 
        return this.serviceId;
    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-engine-api-lib/lib-documentation
    public getIPAddress(): string {
        return this.ipAddress;
    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-engine-api-lib/lib-documentation
    public getSharedDirectory(): SharedPath {
        return  this.sharedDirectory
    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-engine-api-lib/lib-documentation
    public async execCommand(command: string[]): Promise<Result<[number, string], Error>> {
        const serviceId: ServiceID = this.serviceId;
        const args: ExecCommandArgs = newExecCommandArgs(this.enclaveId, serviceId, command);

        const promiseExecCommand: Promise<Result<ExecCommandResponse, Error>> = new Promise((resolve, _unusedReject) => {
            this.client.execCommand(args, (error: grpc.ServiceError | null, response?: ExecCommandResponse) => {
                if (error === null) {
                    if (!response) {
                        resolve(err(new Error("No error was encountered but the response was still falsy; this should never happen")));
                    } else {
                        resolve(ok(response!));
                    }
                } else {
                    resolve(err(error));
                }
            })
        });
        const resultExecCommand: Result<ExecCommandResponse, Error> = await promiseExecCommand;
        if (!resultExecCommand.isOk()) {
            return err(resultExecCommand.error);
        }
        const resp: ExecCommandResponse = resultExecCommand.value;

        return ok([resp.getExitCode(), resp.getLogOutput()]);
    }
}
